### ISSUE TITLE
cdist is very slow if custom weight vector is supplied

### ISSUE THREAD

**Issue by pmavrodiev** [scipy/scipy#13629]
**Is your feature request related to a problem? Please describe.**
Several distance metrics in `cdist `take an optional weight vector by which to scale the input vectors, e.g. `sqeuclidean`. The issue is that supplying such a vector causes significant performance degradation in terms of computational time.

The underlying bottleneck seems to be the result of the data validation done on the weight vector. The function `_validate_vector` in `distance.py` is called every time the `cdist` function is invoked. When `cdist` is used in an optimization problem with potentially many iterations, `_validate_vector` will be called myriads of times, essentially for no good reason.

The work-around I am currently using is to manually re-scale the input vectors and then supply them to `cdist` with the default `None` value for the weight vector. The performance increase is staggering.

**Describe the solution you'd like**
Wouldn't it be the Pythonic way to simply use the supplied weights vector as it is, and if malformatted raise an appropriate exception, so that the user makes sure the right data type is supplied? As it stands currently, validating data types and squashing to required dimensions is just too time consuming in the context of an optimization problem.

**Describe alternatives you've considered**
Manually re-scale the input vectors by the custom weight vector and use `cdist` with no optional arguments.

**Comment by tupui** [comment]
Actually most functions are doing data validation on the inputs. So it is true that in an optimization problem, there would possibly be a lot of unnecessary operations. A possible way to mitigate this would be to have a way to deactivate data validation. I don't see an obvious way to deal with this problem. We could think about some mechanism to globally deactivate all sort of validation, using a defined signature for all validators and mocking could do the trick even if I am not sure about the overhead.

**Comment by pmavrodiev** [comment]
Thank you for your answer @tupui. I wouldn't go as far as proposing a global switch to turn off all data validation. In this particular case, however, I do believe that not doing any data validation at all on the weights is a good default case. 
I think that any exception arising from invalid weights (e.g. NaN/non-numeric values, etc.) will be easy to track down by the user.  If a user has made it that far to require custom weights, they must know what they're doing, consequently can validate the weights they're supplying.

I'd prefer this to having an extra switch that turns off weight validation, simply because it'll complicate the function signature. But that's really a matter of preference. As it stands, though, I do believe that `cdist`  is an unexpected bottleneck in scenarios where it's most needed.

**Comment by tupui** [comment]
Well I played a bit around and this is actually coming from the fact that applying the weights is an operation done on the metric function itself. So it is done for every points, not just once at the end. Hence the huge slowdown...

So here there is nothing about the validation. To mitigate this, we would need to move the weight part after the computation of the metric. This means separating the metric calculation from applying the weights. Doable, but the strategy being different for the metrics, multiple functions have to be created for that.

**Comment by tupui** [comment]
If you use `sqeuclidean`, there is no validator defined in `_METRICS`. It means that the function will be run again and get called point by point... I am actually not sure if this is wanted. I will dig more.

So right now, some extra operations due to the weighting is done per points. Hence the slow down.

**Comment by tupui** [comment]
Ok it's not about the validator sorry. It's due to `_select_weighted_metric`. `sqeuclidean` is changed here into `test_sqeuclidean` and thus the function is called on every points. Still not sure why this is happening.

**Comment by tupui** [comment]
When there is no C version, the test mode is used with the python version which is passed as a callable. For the test version, there is no vectorization and the function is called for every points.

Looking at the whole function, it could probably be refactored. First to not oblige to have a C version in order to act on all points at the same time, and then to separate the tests from here.

cc @rgommers

**Comment by pmavrodiev** [comment]
> If you use `sqeuclidean`, ~there is no validator defined in `_METRICS`~. It means that the function will get called point by point... I am actually not sure if this is wanted. I will dig more.
> 
> So right now, some extra operations due to the weighting is done per points. Hence the slow down.
> 
> If you try `seuclidean` you will see absolutely no slow down as the validation is done once and computation is done for all points at once.

Unfortunately, I do need `sqeuclidean`, because it supports individual weights for each vector dimension. `seuclidean` does not.

**Comment by rgommers** [comment]
> The work-around I am currently using is to manually re-scale the input vectors and then supply them to `cdist` with the default `None` value for the weight vector. The performance increase is staggering.

Could you share some timings, or benchmarking code? Staggering can mean 20% or 20x, would be nice to have a bit more detail.

> Looking at the whole function, it could probably be refactored. First to not oblige to have a C version in order to act on all points at the same time, and then to separate the tests from here.

`spatial.distance` is a bit of a mess indeed. Two large rewrites have been attempted over the years, but none of them was completed. There's a lot of performance to gain, as well as design consistency. I know @peterbell10 has some interest in working on this topic.

**Comment by pmavrodiev** [comment]
> Could you share some timings, or benchmarking code? Staggering can mean 20% or 20x, would be nice to have a bit more detail.

Hi @rgommers Unfortunately, it'd be quite time-consuming for me to post a minimal working code example based off of my actual application. I have definitely measured at least one order of magnitude slow-down with vectors of moderate size, e.g. (20000 x 50). 

I'm sure that any test with similarly sized vectors can show the performance difference with vs without custom weights. 

> `spatial.distance` is a bit of a mess indeed. Two large rewrites have been attempted over the years, but none of them was completed. There's a lot of performance to gain, as well as design consistency.

I'd like to help out the community, but don't have the bandwidth at the moment. Hopefully just reporting issues would help to prioritize internally.

**Comment by peterbell10** [comment]
Here is a straight-forward example:
```python
In [1]: import numpy as np
   ...: from scipy.spatial.distance import cdist
   ...: a = np.random.rand(100, 10)
   ...: w = np.random.rand(10)
   ...: %timeit cdist(a, a, metric='euclidean')
   ...: %timeit cdist(a, a, w=w, metric='euclidean')
337 µs ± 3.77 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
173 ms ± 1.45 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
We can see that introducing weights results in a 3 order of magnitude slow-down. Exactly because it falls-back to calling `'test_euclidean'` individually for each `x[i], y[j]` pair in the distance matrix.
There is a similar performance cliff for `pdist` as well.

I have time to work on this and am interested in basically re-writing `spatial.distance`. Ideally, the distance metric should be implemented once in C++ for weighted and non-weighted, and there would be infrastructure to generalize that automatically to nd-arrays, `cdist` and `pdist` automatically.
